source/state_representation/src/trajectory/CartesianTrajectory.cpp

```diff
@@ -0,0 +1,210 @@
+  try {
+    this->add_points(points, durations);
+  } catch (...) {
+    throw;
+  }
+}
+
+const std::string& CartesianTrajectory::get_reference_frame() const {
+  return this->reference_frame_;
+}
+
+void CartesianTrajectory::set_reference_frame(const CartesianPose& pose) {
+  if (pose.is_empty()) {
+    throw exceptions::EmptyStateException("Pose is empty");
+  } else if (this->get_size() == 0) {
+    throw exceptions::EmptyStateException("Trajectory is empty");
+  }
+  this->reference_frame_ = pose.get_reference_frame();
+  auto points = this->get_points();
+  for (auto& point : points) {
+    point *= pose;
+  }
+  try {
+    this->set_points(points, this->get_durations());
+  } catch (...) {
+    throw;
+  }
+}
+
+void CartesianTrajectory::add_point(const CartesianState& new_point, const std::chrono::nanoseconds& duration) {
+  if (new_point.is_empty()) {
+    throw exceptions::EmptyStateException("The Cartesian state provided is empty");
+  } else if (new_point.get_reference_frame() != this->reference_frame_) {
+    throw exceptions::IncompatibleReferenceFramesException(
+        "Incompatible reference frames: " + new_point.get_reference_frame() + " and " + this->reference_frame_
+    );
+  }
+  CartesianTrajectoryPoint trajectory_point;
+  trajectory_point.data = new_point.data();
+  trajectory_point.duration = duration;
+  trajectory_point.name = new_point.get_name();
+  this->TrajectoryBase<CartesianTrajectoryPoint>::add_point(trajectory_point);
+}
+
+void CartesianTrajectory::add_points(
+    const std::vector<CartesianState>& new_points, const std::vector<std::chrono::nanoseconds>& durations
+) {
+  if (new_points.empty()) {
+    throw exceptions::IncompatibleSizeException("No points provided");
+  } else if (new_points.size() != durations.size()) {
+    throw exceptions::IncompatibleSizeException("The size of the points and durations vectors are not equal");
+  } else if (std::ranges::any_of(new_points, [&](const auto& p) { return p.is_empty(); })) {
+    throw exceptions::EmptyStateException("Vector contains at least one point that is empty");
+  } else if (!std::ranges::all_of(new_points, [&](const auto& p) {
+               return p.get_reference_frame() == this->reference_frame_;
+             })) {
+    throw exceptions::IncompatibleReferenceFramesException("Incompatible reference frames within the points vector");
+  }
+  for (unsigned int i = 0; i < new_points.size(); ++i) {
+    add_point(new_points[i], durations[i]);
+  }
+}
+
+void CartesianTrajectory::insert_point(
+    const CartesianState& new_point, const std::chrono::nanoseconds& duration, unsigned int index
+) {
+  if (new_point.is_empty()) {
+    throw exceptions::EmptyStateException("Point is empty");
+  } else if (new_point.get_reference_frame() != this->reference_frame_) {
+    throw exceptions::IncompatibleReferenceFramesException(
+        "Incompatible reference frames: " + new_point.get_reference_frame() + " and " + this->reference_frame_
+    );
```

Comment:
These lines are also duplicated, which speaks for a static helper up top that does the same
```
assert_valid(const CartesianState& point, const std::string& reference_frame) {
  throw if point empty
  throw if reference frame not the same
}
```